paper/example/ex_mnitst.jl

```diff
@@ -0,0 +1,345 @@
+
+
+"""
+Lenet model
+Data need to be MNIST (#todo test with CIFAR10))
+Returns Knet model and KnetNLPModel model
+"""
+function lenet_prob(xtrn, ytrn, xtst, ytst; minibatchSize = 100)
+
+    LeNet = Chainnll(
+        Conv(5, 5, 1, 20), # Conv(filter_length, filter_width, number of input channels from previous layer :n filters, 3 for RGB, 1 for Grey..., number of filters)
+        Conv(5, 5, 20, 50),
+        Dense(800, 500),
+        Dense(500, 10, identity),
+    )
+
+    LeNetNLPModel = KnetNLPModel(
+        LeNet;
+        data_train = (xtrn, ytrn),
+        data_test = (xtst, ytst),
+        size_minibatch = minibatchSize,
+    )
+
+
+    return LeNet, LeNetNLPModel
+
+end
+
+####################################################
+# R2 supports callback function so we don't need to 
+# init every epoch 
+# currently we have Cb stopping the DNN model after
+# some iterations, but TBD other stopping conditions
+####################################################
+
+function cb(
+    nlp,
+    solver,
+    stats,
+    data::StochasticR2Data,
+)
+    # Max epoch
+    if data.epoch == data.max_epoch
+        stats.status = :user
+        return
+    end
+
+    data.i = KnetNLPModels.minibatch_next_train!(nlp)
+    if data.i == 1   # once one epoch is finished     
+        # reset
+        data.grads_arr = []
+        data.epoch += 1
+        acc = KnetNLPModels.accuracy(nlp) # accracy of the minibatch on the test Data
+        train_acc = Knet.accuracy(nlp.chain; data = nlp.training_minibatch_iterator) #TODO minibatch acc.
+        @info @sprintf "Current epoch:  %d out of max epoch: %d, \t train_acc: %f \t test_acc: %f" data.epoch data.max_epoch train_acc acc
+        append!(data.train_acc_arr, train_acc) #TODO fix this to save the acc
+        append!(data.acc_arr, acc) #TODO fix this to save the acc
+        append!(data.epoch_arr, data.epoch)
+    end
+
+end
+####################################################
+# Training using R2
+####################################################
+
+
+function train_knetNLPmodel!(
+    modelNLP,
+    solver,
+    xtrn,
+    ytrn;
+    mbatch = 64, #128     #TODO see if we need this , in future we can update the number of batch size in different epoch
+    mepoch = 10, # 100
+    verbose = -1,
+    atol::T = √eps(T),
+    rtol::T = √eps(T),
+    η1 = eps(T)^(1 / 4),
+    η2 = T(0.95),
+    γ1 = T(1 / 2),
+    γ2 = 1 / γ1,
+    σmin = zero(T),# change this
+    β::T = T(0),
+    max_time = Inf, 
+) where {T}
+    stochastic_data = StochasticR2Data(0, 0, mepoch, [], [], [], [], atol)
+    solver_stats = solver(
+        modelNLP;
+        atol = atol,
+        rtol = rtol,
+        η1 = η1,
+        η2 = η2,
+        γ1 = γ1,
+        γ2 = γ2,
+        σmin = σmin,
+        β = β,
+        max_time = max_time,
+        verbose = verbose,
+        callback = (nlp, solver, stats) ->
+            cb(nlp, solver, stats, stochastic_data),
+    )
```

Comment:
That way, we only keep 4 of those lines 
```julia
    stochastic_data = StochasticR2Data(0, 0, mepoch, [], [], [], [], atol)
    solver_stats = solver(
        modelNLP;
        callback = (nlp, solver, stats) ->
            cb(nlp, solver, stats, stochastic_data),
```
and maybe `verbose`.

These lines must show the power of combining Solvers and KnetNLPModels inside the JSO ecosystem.
The simplier it is, the best advertising it is (for me).

Reply:
I agree, 
I will update this